internal/app/machined/pkg/system/services/etcd.go

```diff
@@ -600,24 +600,60 @@ func (e *Etcd) recoverFromSnapshot(hostname, primaryAddr string) error {
 func promoteMember(ctx context.Context, r runtime.Runtime, memberID uint64) error {
 	// try to promote a member until it succeeds (call might fail until the member catches up with the leader)
 	// promote member call will fail until member catches up with the master
+	//
+	// iterate over all endpoints until we find the one which works
+	// if we stick with the default behavior, we might hit the member being promoted, and that will never
+	// promote itself.
+	idx := 0
+
 	return retry.Constant(10*time.Minute,
 		retry.WithUnits(15*time.Second),
 		retry.WithJitter(time.Second),
 		retry.WithErrorLogging(true),
 	).RetryWithContext(ctx, func(ctx context.Context) error {
-		client, err := etcd.NewClientFromControlPlaneIPsNoDiscovery(ctx, r.State().V1Alpha2().Resources())
+		endpoints, err := etcd.GetEndpoints(ctx, r.State().V1Alpha2().Resources())
 		if err != nil {
 			return retry.ExpectedError(err)
 		}
 
-		defer client.Close() //nolint:errcheck
+		if len(endpoints) == 0 {
+			return retry.ExpectedErrorf("no endpoints")
+		}
+
+		// try to iterate all available endpoints in the time available for an attempt
+		for i := 0; i < len(endpoints); i++ {
+			select {
+			case <-ctx.Done():
+				return retry.ExpectedError(ctx.Err())
```

Comment:
is this expected, retrying when the context is done?

Reply:
so I was trying to do the following logic here:

* `ctx` in the retry loop  (in `RetryContext`) version is actually set with the deadline of the attempt timeout (i.e. 15 seconds in this case)
* so if I return not expected error, `RetryContext` loop will abort
* what I want to do: during those 15 seconds allocated for an attempt, I want to try as many different etcd endpoints as I can